indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java

```diff
@@ -2950,21 +2959,21 @@ private void checkTaskDuration() throws ExecutionException, InterruptedException
         }
       }
 
-
-      boolean stopTasksEarly = false;
-      if (earlyStopTime != null && (earlyStopTime.isBeforeNow() || earlyStopTime.isEqualNow())) {
-        log.info("Early stop requested - signalling tasks to complete");
-
-        earlyStopTime = null;
-        stopTasksEarly = true;
-      }
-
-
-      // if this task has run longer than the configured duration, signal all tasks in the group to persist
-      if (earliestTaskStart.plus(ioConfig.getTaskDuration()).isBeforeNow() || stopTasksEarly) {
+      if (stopTasksEarly) {
         log.info("Task group [%d] has run for [%s]", groupId, ioConfig.getTaskDuration());
         futureGroupIds.add(groupId);
         futures.add(checkpointTaskGroup(group, true));
+      } else if (earliestTaskStart.plus(ioConfig.getTaskDuration()).isBeforeNow()) {
+        // if this task has run longer than the configured duration
+        // as long as the pending task groups are less than the configured stop task count.
+        if (ioConfig.getStopTaskCount() == ioConfig.getTaskCount()
+            || pendingCompletionTaskGroups.values().stream().mapToInt(CopyOnWriteArrayList::size).sum()
```

Comment:
Doesn't pendingCompletionTaskGroups gets updated outside of this loop? meaning that you would add all tasks to be checkpoint before pendingCompletionTaskGroups gets updated?

Reply:
The `break` on line 2976 makes it so that pendingCompletionTasksGroup is updated, and other tasks are stopped on the next call of `RunNotice#handle`

Reply:
ah! I missed the `break`. Thanks